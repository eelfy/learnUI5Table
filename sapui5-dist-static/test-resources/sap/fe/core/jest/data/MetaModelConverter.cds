using {cuid, managed} from '@sap/cds/common';

namespace sap.fe.test;

entity TestEntity
  @(Common       : {
        SemanticKey                              : [ID]
        }) {
    key ID                                : Integer;
        name                              : String;
        propWithText					  : String @(
		Common        : {
						Text            : propWithTextText,
						TextArrangement : #TextFirst
					});
        propWithTextText                              : String;
        imageURL                          : String      @UI         : {IsImageURL : true};
        imageURLStream                    : String      @(
            Core.MediaType : 'image/jpg',
            Core.IsURL
        );
        mediaType                         : String      @Core.isMediaType;
        imageStream                       : LargeBinary @(
            Core.MediaType : 'image/png',
            Core.Computed
        );
        stream                       		: LargeBinary;
        preferredItem                     : Boolean;
        creationDate                      : Timestamp;
        usageStartTime                    : Time;
        usageCoefficient                  : Double;
        usageCounter                      : Integer64;
        usageProgress                     : Decimal(4, 2);
        usageDuration                     : String      @odata.Type : 'Edm.Duration';
        emailAddress                      : String      @Communication.IsEmailAddress;
        phoneNumber                       : String      @Communication.IsPhoneNumber;
        headerDate                        : Date;
        quantity1                    	  : Integer64	@Measures.Unit : unit1;
        quantity2                         : Decimal(10, 2)	@Measures.Unit : unit2;
        quantityWithStaticUnit			  : Integer64	@Measures.Unit : 'mol';
        amount1							  : Integer64	@Measures.ISOCurrency : unit1;
        unit1                             : String;
		@readonly unit2                   : String;
		multiline						  : String @UI.MultiLineText: true;
}

annotate Items with @(UI : {
    LineItem                           : [
    {Value : ID},
    {Value : name},
    {Value : imageURL},
    {Value : imageURLStream},
    {Value : imageStream},
    {Value : preferredItem},
    {Value : creationDate},
 	{Value: usageProgress},
   	{Value : usageStartTime},
    {Value : usageCoefficient},
    {Value : usageCounter},
    {Value : emailAddress},
    {Value : phoneNumber},
    {Value : headerDate},
    {Value : quantity1},
    {Value : quantity2},
    {Value : amount1},
    {Value : quantityWithStaticUnit}
    ],
    LineItem#bis:[
    {Value : multiline},
    {
                $Type : 'UI.DataFieldWithUrl',
                Value : ID
            }
    ],
    DataPoint #withValueFormat			: {
    	Value: usageProgress,
    	ValueFormat: {
    		NumberOfFractionalDigits: 1
    	},
    	Title: 'Usage Progress (Decimal)'
    },
     FieldGroup #foo   : {Data : [{
            $Type : 'UI.DataField',
            Value : ID
        }]},

},
Common        : {
	SemanticObject                  : 'LineItems'
});




service JestService {
    @odata.draft.enabled
    entity TestEntity      as
        select from test.TestEntity {
        *
        }
}