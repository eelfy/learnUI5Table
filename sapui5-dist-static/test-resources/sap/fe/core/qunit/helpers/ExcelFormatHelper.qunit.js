/* global QUnit */
sap.ui.define(["sap/fe/core/helpers/ExcelFormatHelper"], function(ExcelFormatHelper) {
	"use strict";
	QUnit.module("Excel Format checks on Date/Time/DateTime dataTypes");

	QUnit.test("Get excel format for a Date property from JS Date", function(assert) {
		assert.equal(ExcelFormatHelper.getExcelDatefromJSDate(), "mmm d, yyyy", "Date Format for excel = mmm d, yyyy");
	});
	QUnit.test("Get excel format for a DateTime property from JS DateTime", function(assert) {
		assert.equal(
			ExcelFormatHelper.getExcelDateTimefromJSDateTime(),
			"mmm d, yyyy, h:mm:ss AM/PM",
			"DateTime Format for excel = mmm d, yyyy, h:mm:ss AM/PM"
		);
	});
	QUnit.test("Get excel format for a Time property from JS Time", function(assert) {
		assert.equal(ExcelFormatHelper.getExcelTimefromJSTime(), "h:mm:ss AM/PM", "Time Format for excel = h:mm:ss AM/PM");
	});
});
