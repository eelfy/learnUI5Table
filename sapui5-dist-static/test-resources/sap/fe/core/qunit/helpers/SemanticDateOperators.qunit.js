/* global QUnit */
sap.ui.define(["sap/fe/core/helpers/SemanticDateOperators"], function(SemanticDateOperators) {
	"use strict";

	QUnit.module("SemanticDate Operators", {});

	var aSemanticMappingTestData = [
		{
			desc: "with 'key' and 'equals'",
			operatorConfiguration: [
				{
					"path": "key",
					"equals": "TODAY,YESTERDAY",
					"exclude": false
				}
			],
			expectedOperations: ["TODAY", "YESTERDAY"]
		},
		{
			desc: "with 'key' and 'contains'",
			operatorConfiguration: [
				{
					"path": "key",
					"contains": "QUARTER",
					"exclude": false
				}
			],
			expectedOperations: [
				"THISQUARTER",
				"LASTQUARTER",
				"LASTQUARTERS",
				"NEXTQUARTER",
				"NEXTQUARTERS",
				"QUARTER1",
				"QUARTER2",
				"QUARTER3",
				"QUARTER4"
			]
		},
		{
			desc: "with 'category' and 'equals'",
			operatorConfiguration: [
				{
					"path": "category",
					"equals": "FIXED.DATE",
					"exclude": false
				}
			],
			expectedOperations: ["TODAY", "YESTERDAY", "TOMORROW"]
		},
		{
			desc: "with 'category' and 'contains'",
			operatorConfiguration: [
				{
					"path": "category",
					"contains": "DYNAMIC.DATE",
					"exclude": false
				}
			],
			expectedOperations: ["DATE", "FROM", "TO", "DATERANGE", "TODAYFROMTO", "LASTDAYS", "NEXTDAYS"]
		},
		{
			desc: "with 'key' and 'equals' - exclude: true",
			operatorConfiguration: [
				{
					"path": "key",
					"equals": "TODAY",
					"exclude": true
				}
			],
			expectedOperations: [
				"DATE",
				"FROM",
				"TO",
				"DATERANGE",
				"SPECIFICMONTH",
				"TODAYFROMTO",
				"YESTERDAY",
				"TOMORROW",
				"LASTDAYS",
				"NEXTDAYS",
				"THISWEEK",
				"LASTWEEK",
				"LASTWEEKS",
				"NEXTWEEK",
				"NEXTWEEKS",
				"THISMONTH",
				"LASTMONTH",
				"LASTMONTHS",
				"NEXTMONTH",
				"NEXTMONTHS",
				"THISQUARTER",
				"LASTQUARTER",
				"LASTQUARTERS",
				"NEXTQUARTER",
				"NEXTQUARTERS",
				"QUARTER1",
				"QUARTER2",
				"QUARTER3",
				"QUARTER4",
				"THISYEAR",
				"LASTYEAR",
				"LASTYEARS",
				"NEXTYEAR",
				"NEXTYEARS",
				"YEARTODATE"
			]
		},
		{
			desc: "with 'category' and 'contains' - exclude: true",
			operatorConfiguration: [
				{
					"path": "category",
					"contains": "DYNAMIC",
					"exclude": true
				}
			],
			expectedOperations: [
				"TODAY",
				"YESTERDAY",
				"TOMORROW",
				"THISWEEK",
				"LASTWEEK",
				"NEXTWEEK",
				"THISMONTH",
				"LASTMONTH",
				"NEXTMONTH",
				"THISQUARTER",
				"LASTQUARTER",
				"NEXTQUARTER",
				"QUARTER1",
				"QUARTER2",
				"QUARTER3",
				"QUARTER4",
				"THISYEAR",
				"LASTYEAR",
				"NEXTYEAR",
				"YEARTODATE"
			]
		},
		{
			desc: "with empty filter settings",
			operatorConfiguration: [],
			expectedOperations: []
		},
		{
			desc: "with multiple filter settings",
			operatorConfiguration: [
				{
					"path": "key",
					"equals": "TODAY,LASTYEAR,NEXTYEAR",
					"exclude": true
				},
				{
					"path": "category",
					"contains": "MONTH",
					"exclude": true
				},
				{
					"path": "key",
					"contains": "QUARTER",
					"exclude": true
				}
			],
			expectedOperations: [
				"DATE",
				"FROM",
				"TO",
				"DATERANGE",
				"TODAYFROMTO",
				"YESTERDAY",
				"TOMORROW",
				"LASTDAYS",
				"NEXTDAYS",
				"THISWEEK",
				"LASTWEEK",
				"LASTWEEKS",
				"NEXTWEEK",
				"NEXTWEEKS",
				"THISYEAR",
				"LASTYEARS",
				"NEXTYEARS",
				"YEARTODATE"
			]
		}
	];

	aSemanticMappingTestData.forEach(function(oTestData) {
		QUnit.test("getFilterOperations - " + oTestData.desc, function(assert) {
			var aSemanticDateOps = SemanticDateOperators.getFilterOperations(oTestData.operatorConfiguration);
			assert.equal(
				JSON.stringify(aSemanticDateOps),
				JSON.stringify(oTestData.expectedOperations),
				"testing semantic date settings " + oTestData.desc + " - done"
			);
		});
	});

	var aFilterConditionsTestMapping = [
		{
			desc: "with no semantic date entered in the Filter Bar",
			condition: {
				"$editState": [
					{
						operator: "DRAFT_EDIT_STATE",
						values: ["ALL"]
					}
				],
				LastChangedDateTime: [
					{
						operator: "EQ",
						values: ["2020-12-03T15:32:05+0530"]
					}
				]
			},
			expectedResult: true
		},
		{
			desc: "with semantic date entered in the Filter Bar",
			condition: {
				"$editState": [
					{
						operator: "DRAFT_EDIT_STATE",
						values: ["ALL"]
					}
				],
				LastChangedDateTime: [
					{
						operator: "EQ",
						values: ["2020-12-03T15:32:05+0530"]
					}
				],
				SalesOrderDate: [
					{
						operator: "YESTERDAY",
						values: ["Dec 2, 2020"]
					}
				]
			},
			expectedResult: false
		}
	];
	aFilterConditionsTestMapping.forEach(function(oTestData) {
		QUnit.test("hasSemanticDateOperations - " + oTestData.desc, function(assert) {
			var bHasSemanticDateOperation = SemanticDateOperators.hasSemanticDateOperations(oTestData.condition);
			assert.equal(
				bHasSemanticDateOperation,
				oTestData.expectedResult,
				"testing hasSemanticDateOperations " + oTestData.desc + " - done"
			);
		});
	});
});
