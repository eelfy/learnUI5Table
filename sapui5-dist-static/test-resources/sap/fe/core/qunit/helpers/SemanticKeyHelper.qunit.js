/* global QUnit */
sap.ui.define(["sap/ui/model/json/JSONModel", "sap/fe/core/helpers/SemanticKeyHelper"], function(JSONModel, SemanticKeyHelper) {
	"use strict";

	var oMetaModel = new JSONModel({
		"SalesOrderManage": {
			"$kind": "EntitySet",
			"$Type": "com.c_salesordermanage_sd.SalesOrderManage"
		},

		"com.c_salesordermanage_sd.SalesOrderManage": {
			"$Key": ["ID", "IsActiveEntity"],
			"ID": {
				"$Type": "Edm.Guid"
			},
			"SalesOrder": {
				"$Type": "Edm.String"
			},
			"IsActiveEntity": {
				"$Type": "Edm.Boolean"
			}
		}
	});

	function createContextStub(sPath, sEntitySetName, mProperties, aSemanticKeys) {
		oMetaModel.getMetaContext = function() {
			return {
				getObject: function() {
					return sEntitySetName;
				}
			};
		};
		oMetaModel.setProperty("/" + sEntitySetName + "/@com.sap.vocabularies.Common.v1.SemanticKey", aSemanticKeys);

		return {
			isA: function() {
				return false;
			},
			getPath: function() {
				return sPath;
			},
			getModel: function() {
				return {
					getMetaModel: function() {
						return oMetaModel;
					}
				};
			},
			getProperty: function(sKey) {
				return mProperties[sKey];
			}
		};
	}

	QUnit.module("SemanticKey helper", {});

	var aSemanticMappingTestData = [
		{
			desc: "One semantic key",
			path: "/SalesOrderManage(ID=123, IsActiveEntity=true)",
			entitySet: "SalesOrderManage",
			contextKeys: { ID: 123 },
			semanticKeys: ["ID"],
			expectedSemPath: "/SalesOrderManage(123)"
		},
		{
			desc: "Two semantic keys",
			path: "/SalesOrderManage(ID=123, IsActiveEntity=true)",
			entitySet: "SalesOrderManage",
			contextKeys: { ID: 123, IsActiveEntity: true },
			semanticKeys: ["ID", "IsActiveEntity"],
			expectedSemPath: "/SalesOrderManage(ID=123,IsActiveEntity=true)"
		},
		{
			desc: "One semantic key (string)",
			path: "/SalesOrderManage(ID=123, IsActiveEntity=true)",
			entitySet: "SalesOrderManage",
			contextKeys: { SalesOrder: "ABCD" },
			semanticKeys: ["SalesOrder"],
			expectedSemPath: "/SalesOrderManage('ABCD')"
		},
		{
			desc: "Two semantic keys (string)",
			path: "/SalesOrderManage(ID=123, IsActiveEntity=true)",
			entitySet: "SalesOrderManage",
			contextKeys: { ID: 123, SalesOrder: "ABCD" },
			semanticKeys: ["ID", "SalesOrder"],
			expectedSemPath: "/SalesOrderManage(ID=123,SalesOrder='ABCD')"
		},
		{
			desc: "One semantic key (not found in context)",
			path: "/SalesOrderManage(ID=123, IsActiveEntity=true)",
			entitySet: "SalesOrderManage",
			contextKeys: {},
			semanticKeys: ["SalesOrder"],
			expectedSemPath: "/SalesOrderManage(ID=123, IsActiveEntity=true)"
		},
		{
			desc: "One semantic key (found in context but not set/available)",
			path: "/SalesOrderManage(ID=123, IsActiveEntity=true)",
			entitySet: "SalesOrderManage",
			contextKeys: { SalesOrder: null },
			semanticKeys: ["SalesOrder"],
			expectedSemPath: "/SalesOrderManage(ID=123, IsActiveEntity=true)"
		},
		{
			desc: "Two semantic keys (not found in context)",
			path: "/SalesOrderManage(ID=123, IsActiveEntity=true)",
			entitySet: "SalesOrderManage",
			contextKeys: { ID: 123 },
			semanticKeys: ["ID", "SalesOrder"],
			expectedSemPath: "/SalesOrderManage(ID=123, IsActiveEntity=true)"
		},
		{
			desc: "Path not eligible for semantic key mapping",
			path: "/SalesOrderManage(ID=123, IsActiveEntity=true)/_Item(XX)",
			entitySet: "_Item",
			contextKeys: { ID: 123 },
			semanticKeys: ["ID"],
			expectedSemPath: "/SalesOrderManage(ID=123, IsActiveEntity=true)/_Item(XX)"
		},
		{
			desc: "No semantic key",
			path: "/SalesOrderManage(ID=123, IsActiveEntity=true)",
			entitySet: "SalesOrderManage",
			contextKeys: { ID: 123 },
			semanticKeys: null,
			expectedSemPath: "/SalesOrderManage(ID=123, IsActiveEntity=true)"
		}
	];

	aSemanticMappingTestData.forEach(function(oTestData) {
		QUnit.test("getSemanticPath - " + oTestData.desc, function(assert) {
			var aSemanticKeys =
				oTestData.semanticKeys &&
				oTestData.semanticKeys.map(function(key) {
					return { $PropertyPath: key };
				});
			var oContext = createContextStub(oTestData.path, oTestData.entitySet, oTestData.contextKeys, aSemanticKeys);

			var sSemPath = SemanticKeyHelper.getSemanticPath(oContext);
			//assert.equal(oMapping.technicalPath, oTestData.path, "testing technical path value");
			assert.equal(sSemPath, oTestData.expectedSemPath, "testing semantic path value");
		});
	});
});
