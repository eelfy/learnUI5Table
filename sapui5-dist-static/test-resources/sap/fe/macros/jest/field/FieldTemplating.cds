using {cuid, managed} from '@sap/cds/common';

namespace sap.fe.test;

entity getBindingWithTextArrangement {
	key ID                       : Integer;
	noText		                 : String;
	textOnly	                 : String @(
                                          	Common : {
                                          		Text            : targetText,
                                          		TextArrangement : #TextOnly
                                          	});
	textFirst		             : String @(
											Common : {
												Text            : targetText,
												TextArrangement : #TextFirst
											});
	textLast	                 : String @(
											Common : {
												Text            : targetText,
												TextArrangement : #TextLast
											});
	textSeparate                 : String @(
											Common : {
												Text            : targetText,
												TextArrangement : #TextSeparate
											});
	targetText	                 : String;
}

service JestService {
    @odata.draft.enabled
    entity getBindingWithTextArrangement      as
        select from test.getBindingWithTextArrangement {
        *
        }
}