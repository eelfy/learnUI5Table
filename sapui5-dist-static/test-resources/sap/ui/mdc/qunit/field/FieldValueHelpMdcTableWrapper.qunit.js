// Use this test page to test the API and features of the FieldHelp.
// The interaction with the Field is tested on the field test page.

/* global QUnit */
/*eslint max-nested-callbacks: [2, 6]*/

sap.ui.define([
	"sap/ui/mdc/field/FieldValueHelpMdcTableWrapper",
	"sap/ui/mdc/field/FieldValueHelpUITableWrapper",
	"sap/ui/mdc/field/FieldValueHelpMTableWrapper",
	"sap/ui/mdc/field/FieldValueHelpDelegate",
	"sap/ui/mdc/field/InParameter",
	"sap/ui/mdc/field/OutParameter",
	"sap/ui/mdc/Table",
	"sap/ui/mdc/table/Column",
	"sap/m/Text",
	"sap/ui/model/json/JSONModel"
], function (
		FieldValueHelpMdcTableWrapper,
		FieldValueHelpUITableWrapper,
		FieldValueHelpMTableWrapper,
		FieldValueHelpDelegate,
		InParameter,
		OutParameter,
		Table,
		Column,
		Text,
		JSONModel
	) {
	"use strict";


	var oModel;


	var oWrapper;
	var oTable;
	var oFilterBar;
	var iSelect = 0;
	var iNavigate = 0;
	var iDataUpdate = 0;

	var _mySelectionChangeHandler = function(oEvent) {
		iSelect++;
	};

	var _myNavigateHandler = function(oEvent) {
		iNavigate++;
	};

	var _myDataUpdateHandler = function(oEvent) {
		iDataUpdate++;
	};

	/* use dummy VieldValueHelp just to test API */
//	var bSingleSelection = true;
	var sKeyPath = "key";
	var sDescriptionPath = "text";
	var iMaxConditions = -1;
	var bUseInParameters = false;
	var oInParameter = new InParameter({helpPath: "additionalText"});
	var bUseOutParameters = false;
	var oOutParameter = new OutParameter({helpPath: "additionalText"});
	var oValueHelp = {
//			_getSingleSelection: function() {
//				return bSingleSelection;
//			},
			_getKeyPath: function() {
				return sKeyPath;
			},
			getDescriptionPath: function() {
				return sDescriptionPath;
			},
			getMaxConditions: function() {
				return iMaxConditions;
			},
			isA: function(sName) {
				return sName === "sap.ui.mdc.field.FieldValueHelp" ? true : false;
			},
			getModel: function(sName) {
				return oModel;
			},
			invalidate: function(oOrigin) {},
			getInParameters: function() {
				if (bUseInParameters) {
					return [oInParameter];
				} else {
					return [];
				}
			},
			getOutParameters: function() {
				if (bUseOutParameters) {
					return [oOutParameter];
				} else {
					return [];
				}
			},
			getDelegate: function() {
				return {name: "sap/ui/mdc/field/FieldValueHelpDelegate", payload: {}};
			},
			getControlDelegate: function () {
				return FieldValueHelpDelegate;
			},
			getPayload: function() {
				return {};
			},
			getFilterBar: function () {
				oFilterBar = oFilterBar || new sap.ui.mdc.filterbar.vh.FilterBar({
					liveMode: false,
					delegate: {name: 'sap/ui/mdc/filterbar/vh/GenericFilterBarDelegate', payload: {collectionName: ''}},
					filterItems: [
						new sap.ui.mdc.FilterField({
							delegate: {name: 'sap/ui/mdc/odata/v4/FieldBaseDelegate', payload: {}},
							conditions: "{$filters>/conditions/text}"
						})
					],
					basicSearchField: new sap.ui.mdc.FilterField({
						delegate: {name: 'sap/ui/mdc/odata/v4/FieldBaseDelegate', payload: {}},
						dataType: "Edm.String",
						conditions: "{$filters>/conditions/$search}",
						width: "50%",
						maxConditions: 1,
						placeholder: "Search"
					})
				});

				return oFilterBar;
			}
	};

	var oClock;

	var _initTable = function (bFVH, oTableType) {
		oTable = new Table("T1", {
			header: "",
			showRowCount: true,
			width: "26rem",
			type: oTableType,
			delegate: {
				name: "sap/ui/mdc/TableDelegate",
				payload: {
					collectionPath: "/items"
				}
			},
			columns: [
				new Column({dataProperty:"key", template: new Text({text: "{key}"})}),
				new Column({dataProperty:"text", template: new Text({text: "{text}"})}),
				new Column({dataProperty:"additionalText", template: new Text({text: "{additionalText}"})})
			]
		});

		if (!bFVH) {
			oTable.setModel(oModel); // as ValueHelp is faked
		}

		oWrapper = new FieldValueHelpMdcTableWrapper("W1", {
					selectedItems: [{key: "I2", description: "Item 2"}],
					selectionChange: _mySelectionChangeHandler,
					navigate: _myNavigateHandler,
					dataUpdate: _myDataUpdateHandler
				});
		if (!bFVH) {
			oWrapper.getParent = function() {
				return oValueHelp;
			};
		}
		oWrapper.setTable(oTable);
	};

	var _initWrapper = function(bFVH) {
		oModel = new JSONModel({
			items:[{text: "Item 1", key: "I1", additionalText: "Text 1"},
				   {text: "Item 2", key: "I2", additionalText: "Text 2"},
				   {text: "X-Item 3", key: "I3", additionalText: "Text 3"}]
			});
		sap.ui.getCore().setModel(oModel);
	};

	var _teardown = function() {
		if (oClock) {
			oClock.restore();
			oClock = undefined;
		}
		oTable.destroy();
		oTable = undefined;
		oWrapper.destroy();
		oWrapper = undefined;
		iSelect = 0;
		iNavigate = 0;
		iDataUpdate = 0;
		sKeyPath = "key";
		sDescriptionPath = "text";
		iMaxConditions = -1;
		bUseInParameters = false;
		bUseOutParameters = false;
		//FieldValueHelpMdcTableWrapper._init();
		oModel.destroy();
		oModel = undefined;

		if (oFilterBar) {
			oFilterBar.destroy();
			oFilterBar = null;
		}
	};

	QUnit.module("API", {
		beforeEach: function() {
			_initWrapper(false);
			},
		afterEach: _teardown
	});


	QUnit.test("initialize grid table", function(assert) {
		var fnDone = assert.async();

		_initTable(false, new sap.ui.mdc.table.GridTableType({rowCountMode: "Fixed"}));

		oWrapper.initialize();

		var oContent = oWrapper.getTable();
		assert.equal(oContent, oTable, "oWrapper content");

		oWrapper._oInnerWrapperClassPromise.then(function () {
			assert.ok(oWrapper.OInnerWrapperClass === FieldValueHelpUITableWrapper, "oWrapper inner class");
			fnDone();
		});
	});

	QUnit.test("initialize responsive table", function(assert) {
		var fnDone = assert.async();

		_initTable(false, new sap.ui.mdc.table.ResponsiveTableType());

		oWrapper.initialize();

		var oContent = oWrapper.getTable();
		assert.equal(oContent, oTable, "oWrapper content");

		oWrapper._oInnerWrapperClassPromise.then(function () {
			assert.ok(oWrapper.OInnerWrapperClass === FieldValueHelpMTableWrapper, "oWrapper inner class");
			fnDone();
		});
	});
});




