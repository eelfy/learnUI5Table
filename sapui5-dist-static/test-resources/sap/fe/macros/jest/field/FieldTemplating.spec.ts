import { getBindingWithTextArrangement, addTextArrangementToBindingExpression } from "sap/fe/macros/field/FieldTemplating";
import path from "path";
import { compileCDS, getMetaModel, getDataModelObjectPathForProperty, evaluateBindingWithModel } from "sap/fe/test/JestTemplatingHelper";
import { convertTypes } from "sap/fe/core/converters/MetaModelConverter";
import { EntitySet, EntityType, Property } from "@sap-ux/annotation-converter";
import { bindingExpression, compileBinding, ifElse } from "sap/fe/core/helpers/BindingExpression";

describe("Field Templating Testing", () => {
	let convertedTypes;
	let entitySet: EntitySet;
	let entityType: EntityType;
	const properties: Record<string, Property> = {};
	beforeAll(async () => {
		const sMetadataUrl = compileCDS(path.join(__dirname, "./FieldTemplating.cds"));
		const metaModel = await getMetaModel(sMetadataUrl);
		convertedTypes = convertTypes(metaModel);
		const maybeEntitySet = convertedTypes.entitySets.find(et => et.name === "getBindingWithTextArrangement");
		if (maybeEntitySet) {
			entitySet = maybeEntitySet;
			entityType = maybeEntitySet.entityType;
			entityType.entityProperties.forEach(prop => {
				properties[prop.name] = prop;
			});
		} else {
			throw new Error("Error in test setup");
		}
	});

	it("#getBindingWithTextArrangement", () => {
		// Cases to consider

		// TextOnly
		const textOnlyBinding = getBindingWithTextArrangement(
			getDataModelObjectPathForProperty(entitySet, properties["textOnly"]),
			bindingExpression("textOnly")
		);
		expect(textOnlyBinding).toMatchSnapshot();
		expect(compileBinding(textOnlyBinding)).toContain("targetText");
		expect(evaluateBindingWithModel(compileBinding(textOnlyBinding), { targetText: "Yolo" })).toEqual("Yolo");

		// No text arrangement
		const noTextBinding = getBindingWithTextArrangement(
			getDataModelObjectPathForProperty(entitySet, properties["noText"]),
			bindingExpression("noText")
		);
		expect(noTextBinding).toMatchSnapshot();
		expect(compileBinding(noTextBinding)).toContain("noText");
		expect(evaluateBindingWithModel(compileBinding(noTextBinding), { noText: "Yolo" })).toEqual("Yolo");

		// TextFirst
		const textFirstBinding = getBindingWithTextArrangement(
			getDataModelObjectPathForProperty(entitySet, properties["textFirst"]),
			bindingExpression("textFirst")
		);
		expect(textFirstBinding).toMatchSnapshot();
		expect(compileBinding(textFirstBinding)).toContain("textFirst");
		expect(compileBinding(textFirstBinding)).toContain("targetText");
		expect(
			evaluateBindingWithModel(compileBinding(textFirstBinding), { textFirst: "Yolo Last", targetText: "Yolo Description" })
		).toEqual("Yolo Description (Yolo Last)");

		// TextLast
		const textLastBinding = getBindingWithTextArrangement(
			getDataModelObjectPathForProperty(entitySet, properties["textLast"]),
			bindingExpression("textLast")
		);
		expect(textLastBinding).toMatchSnapshot();
		expect(compileBinding(textLastBinding)).toContain("textLast");
		expect(compileBinding(textLastBinding)).toContain("targetText");
		expect(
			evaluateBindingWithModel(compileBinding(textLastBinding), { textLast: "Yolo First", targetText: "Yolo Description" })
		).toEqual("Yolo First (Yolo Description)");
	});

	it("#addTextArrangementToBindingExpression", () => {
		const fullContext = getDataModelObjectPathForProperty(entitySet);

		const otherModelBinding = addTextArrangementToBindingExpression(bindingExpression("textLast", "otherModel"), fullContext);
		expect(otherModelBinding).toMatchSnapshot();
		expect(compileBinding(otherModelBinding)).toContain("textLast");
		expect(compileBinding(otherModelBinding)).not.toContain("targetText");

		const addTextLastBinding = addTextArrangementToBindingExpression(bindingExpression("textLast"), fullContext);
		expect(addTextLastBinding).toMatchSnapshot();
		expect(compileBinding(addTextLastBinding)).toContain("textLast");
		expect(compileBinding(addTextLastBinding)).toContain("targetText");
		expect(
			evaluateBindingWithModel(compileBinding(addTextLastBinding), { textLast: "Yolo First", targetText: "Yolo Description" })
		).toEqual("Yolo First (Yolo Description)");

		const complexTextBinding = addTextArrangementToBindingExpression(
			ifElse(bindingExpression("testProp"), bindingExpression("textLast"), bindingExpression("noText")),
			fullContext
		);
		expect(complexTextBinding).toMatchSnapshot();
		expect(compileBinding(complexTextBinding)).toContain("textLast");
		expect(compileBinding(complexTextBinding)).toContain("targetText");
		const modelContent = {
			testProp: true,
			textLast: "Yolo First",
			targetText: "Yolo Description",
			noText: "Not Yolo"
		};
		expect(evaluateBindingWithModel(compileBinding(complexTextBinding), modelContent)).toEqual("Yolo First (Yolo Description)");

		modelContent.testProp = false;
		expect(evaluateBindingWithModel(compileBinding(complexTextBinding), modelContent)).toEqual("Not Yolo");
	});
});
