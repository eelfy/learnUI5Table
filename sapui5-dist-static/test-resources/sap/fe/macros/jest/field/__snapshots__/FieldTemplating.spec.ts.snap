// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field Templating Testing #addTextArrangementToBindingExpression 1`] = `
Object {
  "_type": "Binding",
  "modelName": "otherModel",
  "path": "textLast",
}
`;

exports[`Field Templating Testing #addTextArrangementToBindingExpression 2`] = `
Object {
  "_type": "Formatter",
  "fn": "sap.fe.core.formatters.ValueFormatter",
  "parameters": Array [
    Object {
      "_type": "Constant",
      "value": "formatWithBrackets",
    },
    Object {
      "_type": "Binding",
      "constraints": Object {},
      "formatOptions": Object {
        "parseKeepsEmptyString": true,
      },
      "modelName": undefined,
      "path": "textLast",
      "type": "sap.ui.model.odata.type.String",
    },
    Object {
      "_type": "Binding",
      "modelName": undefined,
      "path": "targetText",
    },
  ],
}
`;

exports[`Field Templating Testing #addTextArrangementToBindingExpression 3`] = `
Object {
  "_type": "IfElse",
  "condition": Object {
    "_type": "Binding",
    "modelName": undefined,
    "path": "testProp",
  },
  "onFalse": Object {
    "_type": "Binding",
    "constraints": Object {},
    "formatOptions": Object {
      "parseKeepsEmptyString": true,
    },
    "modelName": undefined,
    "path": "noText",
    "type": "sap.ui.model.odata.type.String",
  },
  "onTrue": Object {
    "_type": "Formatter",
    "fn": "sap.fe.core.formatters.ValueFormatter",
    "parameters": Array [
      Object {
        "_type": "Constant",
        "value": "formatWithBrackets",
      },
      Object {
        "_type": "Binding",
        "constraints": Object {},
        "formatOptions": Object {
          "parseKeepsEmptyString": true,
        },
        "modelName": undefined,
        "path": "textLast",
        "type": "sap.ui.model.odata.type.String",
      },
      Object {
        "_type": "Binding",
        "modelName": undefined,
        "path": "targetText",
      },
    ],
  },
}
`;

exports[`Field Templating Testing #getBindingWithTextArrangement 1`] = `
Object {
  "_type": "Binding",
  "modelName": undefined,
  "path": "targetText",
}
`;

exports[`Field Templating Testing #getBindingWithTextArrangement 2`] = `
Object {
  "_type": "Binding",
  "constraints": Object {},
  "formatOptions": Object {
    "parseKeepsEmptyString": true,
  },
  "modelName": undefined,
  "path": "noText",
  "type": "sap.ui.model.odata.type.String",
}
`;

exports[`Field Templating Testing #getBindingWithTextArrangement 3`] = `
Object {
  "_type": "Formatter",
  "fn": "sap.fe.core.formatters.ValueFormatter",
  "parameters": Array [
    Object {
      "_type": "Constant",
      "value": "formatWithBrackets",
    },
    Object {
      "_type": "Binding",
      "modelName": undefined,
      "path": "targetText",
    },
    Object {
      "_type": "Binding",
      "constraints": Object {},
      "formatOptions": Object {
        "parseKeepsEmptyString": true,
      },
      "modelName": undefined,
      "path": "textFirst",
      "type": "sap.ui.model.odata.type.String",
    },
  ],
}
`;

exports[`Field Templating Testing #getBindingWithTextArrangement 4`] = `
Object {
  "_type": "Formatter",
  "fn": "sap.fe.core.formatters.ValueFormatter",
  "parameters": Array [
    Object {
      "_type": "Constant",
      "value": "formatWithBrackets",
    },
    Object {
      "_type": "Binding",
      "constraints": Object {},
      "formatOptions": Object {
        "parseKeepsEmptyString": true,
      },
      "modelName": undefined,
      "path": "textLast",
      "type": "sap.ui.model.odata.type.String",
    },
    Object {
      "_type": "Binding",
      "modelName": undefined,
      "path": "targetText",
    },
  ],
}
`;
